framework/src/problems/FEProblemBase.C

```diff
@@ -5852,6 +5854,21 @@ FEProblemBase::setException(const std::string & message)
   _exception_message = message;
 }
 
+void
+FEProblemBase::setException(const ExecFlagType & type, const std::string & message)
+{
+  if (type == EXEC_LINEAR || type == EXEC_NONLINEAR)
```

Comment:
So thinking again about this, I do think we should limit this to `EXEC_LINEAR`. This is because our stopping mechanism in `NonlinearSystem::stopSolve()` is `SNESSetFunctionDomainError(solver.snes());` which is not going to be checked by PETSc during a nonlinear/Jacobian evaluation